(PUP-10420) Allow state machine to retry ssl lock

joshcooper opened the pull request:
Previously, if the state machine failed to acquired the ssl lock it would raise
a Puppet::Error with the message "Another puppet instance is already running;
exiting" causing puppet to exit.

This commit makes it possible for the state machine to retry the ssl lock up to
a maximum deadline. By default, the state machine won't retry.

If we fail to acquire the lock (after retrying or not), we raise Puppet::Error
and state the reason why we're exiting.

joshcooper commented on spec/unit/ssl/state_machine_spec.rb:
```diff
@@ -262,6 +255,58 @@ def expect_lockfile_to_contain(pid)
+      end
+
+      it 'waits for the lock indefinitely by default' do
+        pending("waiting for setting")
```
This is pending

joshcooper commented on lib/puppet/ssl/state_machine.rb:
```diff
@@ -340,6 +380,8 @@ class Done < SSLState; end
   def initialize(waitforcert: Puppet[:waitforcert],
                  maxwaitforcert: Puppet[:maxwaitforcert],
+                 waitforlock: 0, #Puppet[:waitforlock],
+                 maxwaitforlock: 5, #Puppet[:maxwaitforlock],
```
These are pending

puppetcla commented:
CLA signed by all contributors.

melissa commented on spec/unit/ssl/state_machine_spec.rb:
```diff
@@ -262,6 +255,58 @@ def expect_lockfile_to_contain(pid)
+        }.to raise_error(Puppet::Error, /Another puppet instance is already running and the maxwaitforlock timeout has been exceeded; exiting/)
+      end
+
+      it 'waits for the lock indefinitely by default' do
```
Is this consistent with current behavior? It seems like it doesn't wait for the lock at all, just bails if it can't get it?

joshcooper replied:
ah right, "by default" is incorrect, I'll update the description

melissa replied:
Cool, that makes sense that it was the title throwing me off. By default, `waitforlock` is set to 0, so it'll error out if it can't obtain the lock, just like it did previously, yes?

melissa commented:
jenkins please test this with servertests

joshcooper commented:
PUP-8051 has been merged so I rebased this on master. We also added a change so that puppet unlocks the ssl_lock file while waiting for the cert to be signed. So the chances of two instances contending for the same ssl_lock file is reduced, but still possible. To see that, run this in one console:

```ruby
# this is necessary so the pid in the lock file matches a process named "puppet"
$0 = 'puppet'
require 'puppet'
Puppet.initialize_settings
Puppet[:certname] = 'onetwothree'
lockfile = Puppet::Util::Pidlock.new(Puppet[:ssl_lockfile])
lockfile.lock
sleep(1000)
```

Then in another console run:

```
$ bundle exec  puppet ssl bootstrap -d --certname onetwothree --waitforlock 1 --maxwaitforlock 5
...
Debug: Finishing transaction 70256989402560
Debug: Executing: 'ps -p 48847 -o comm='
Debug: Executing: 'ps -p 48847 -o args='
Info: Another puppet instance is already running; waiting for it to finish
Info: Will try again in 1 seconds.
Debug: Executing: 'ps -p 48847 -o comm='
Debug: Executing: 'ps -p 48847 -o args='
Info: Another puppet instance is already running; waiting for it to finish
Info: Will try again in 1 seconds.
Debug: Executing: 'ps -p 48847 -o comm='
Debug: Executing: 'ps -p 48847 -o args='
Info: Another puppet instance is already running; waiting for it to finish
Info: Will try again in 1 seconds.
Debug: Executing: 'ps -p 48847 -o comm='
Debug: Executing: 'ps -p 48847 -o args='
Info: Another puppet instance is already running; waiting for it to finish
Info: Will try again in 1 seconds.
Debug: Executing: 'ps -p 48847 -o comm='
Debug: Executing: 'ps -p 48847 -o args='
Info: Another puppet instance is already running; waiting for it to finish
Info: Will try again in 1 seconds.
Debug: Executing: 'ps -p 48847 -o comm='
Debug: Executing: 'ps -p 48847 -o args='
Debug: Executing: 'ps -p 48847 -o comm='
Debug: Executing: 'ps -p 48847 -o args='
Error: Could not run: Another puppet instance is already running and the maxwaitforlock timeout has been exceeded; exiting
```

joshcooper commented:
jenkins please test this with servertests on redhat7-64a,ubuntu2004-64a,windows2019-64a
